src/Workspaces/Remote/Core/AbstractAssetProvider.cs

```diff
@@ -188,14 +197,14 @@ internal static class AbstractAssetProviderExtensions
         using var _1 = PooledHashSet<Checksum>.GetInstance(out var checksumSet);
         checksumSet.AddAll(checksums.Children);
 
-        using var _2 = ArrayBuilder<T>.GetInstance(checksumSet.Count, out var builder);
+        var builder = ImmutableArray.CreateBuilder<T>(checksumSet.Count);
 
-        await assetProvider.GetAssetsAsync<T, ArrayBuilder<T>>(
+        await assetProvider.GetAssetHelper<T>().GetAssetsAsync(
             assetPath, checksumSet,
             static (checksum, asset, builder) => builder.Add(asset),
             builder,
             cancellationToken).ConfigureAwait(false);
 
-        return builder.ToImmutableAndClear();
+        return builder.MoveToImmutable();
```

Comment:
@ToddGrun note the movetoimmutable here.  the builder itself produces no internal garbage.  it just holds onto the array and count.  MoveToImmtuable is just an ownership transform of its array to the IA result.  So the only thing left is the builder itself, which should be trivially collected.  

Note: one pattern we could consider here is pooling these builders *when* we've successfully moved their internal arrays out.